Add HTTP specification (#799)
ballerina-platform/module-ballerina-http: docs/spec/spec.md

There are few API level additions when it comes to the HTTP/2 design such as Push promise and promise response.
#### 10.1.1. Push Promise and Promise Response
Push Promise and Promise response are the only application level new semantics which are introduced by HTTP2.

Other protocol changes such as streams, messages, frames, request prioritization, flow control, header compression, 
etc. are all lower level changes that can be handled by the HTTP listener seamlessly from the user.

### 10.2. Websocket

The Websocket protocol is also considered as an upgrade from HTTP where the initial handshake is happening over HTTP. 
Therefore websocket applications can use the same http:Listener and the respective port to initialize the websocket 
service.

The WebSocket listener can be constructed with a port or an http:Listener. When a port is passed to the listener 
constructor, the listener opens the port and attaches the upgrader service at the given service path.
When the http:Listener is passed in, the upgrader service gets attached to it at the given service path. The service 
by default attaches to “/”.

```ballerina
http:Listener hl = new(9090);

service /ws on new ws:Listener(hl) {

>     remote function onUpgrade(http:Caller hc, http:Request req) returns ws:Service|ws:UpgradeError {

Reviewer: ```suggestion
    resource function get .(http:Request req) returns websocket:Service|websocket:UpgradeError {
```

        return new MyWSService();
    }
}

service class MyWSService {

>     *ws:Service;
>     remote function onText(ws:Caller caller, string text) {

Reviewer: ```suggestion
    *websocket:Service;
    remote function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
```

>     *ws:Service;
>     remote function onText(ws:Caller caller, string text) {
>         caller->pushText(text);

Reviewer: ```suggestion
    *websocket:Service;
    remote function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
        check caller->writeTextMessage(data);
```

            * 9.1.1.7. Client - Self Signed JWT Auth
            * 9.1.1.8. Client - OAuth2
        * 9.1.2 Imperative Approach
            * 9.1.2.1. Service - File User Store
            * 9.1.2.2. Service - LDAP User Store
            * 9.1.2.3. Service - JWT
            * 9.1.2.4. Service - OAuth2
            * 9.1.2.5. Client - Basic Auth
            * 9.1.2.6. Client - Bearer Token Auth
            * 9.1.2.7. Client - Self Signed JWT Auth
            * 9.1.2.8. Client - OAuth2   
10. [Protocol-upgrade](#10-protocol-upgrade)
    * 10.1. [HTTP2](#101-http2)
        * 10.1.1. [Push promise and promise response](#1011-push-promise-and-promise-response)
    * 10.2. [Websocket](#102-websocket)

## 1. Overview
This specification elaborates on Basic Auth authentication and authorization for all the Ballerina listeners and
clients. The HTTP, gRPC, GraphQL, WebSocket, WebSub protocol-based listeners and clients are secured according to this
specification.

The ballerina language provides the first class support for writing network oriented programming models in a more 
initiative manner. The ballerina-http standard library consumes the rich language construct and creates the 

> programming model to write services.

Reviewer 2: I feel the start can be improved. How about something like this ?

The HTTP standard library is designed to work with HTTP protocol. It includes high level abstractions such as `http:Request`, `http:Response`, `http:Service` and `http:Client` which allow users to consume and produce HTTP services. Further the users can use this library to build other libraries. In fact, the standard libraries such as GraphQL, Websub and WebSubHub use this library internally.  

In addition to functional requirements, this library deals with none functional requirements such as security, observerbility and resiliency. Each requirement is discussed in detail in the coming sections.

## 2. Components
### 2.1. Listener
The HTTP listener object receives network data from a remote process according to the HTTP transport protocol and 
translates the received data into invocations on the resources functions of services that have been 
attached to the listener object. The listener provides the interface between network and services.

As defined in [Ballerina 2021R1 Section 5.7.4](https://ballerina.io/spec/lang/2021R1/#section_5.7.4) the Listener has 
the object constructor and life cycle methods such as attach(), detach(), 'start(), gracefulStop(), and immediateStop().

> #### 2.1.1. Auto binding

Reviewer 2: I feel `binding` is not the correct word here.

Reviewer 2: 2.1.1. Automatically starting the service
2.1.2. Programmatically starting the service

Reviewer 2: Above two might be a better option.

If a service is attached to the listener, then the listener starts listening on the given port after executing 
attach() and start() methods. HTTP listener can be declared as follows honoring to the generic 
[listener declaration](https://ballerina.io/spec/lang/2021R1/#section_8.3.1)

```ballerina
// Listener object constructor
listener http:Listener serviceListener = new(9090, config = {httpVersion: "2.0"});
```
```ballerina
// Service attaches to the Listener
service http:Service /sessionTest on serviceListener {

> 

Reviewer 2: Maybe we can add the same.

`resource function get sayHello(http:Caller caller) {}`

}
```

#### 2.1.2. Late binding

Users can programmatically start the listener by calling each lifecycle method as follows.

```ballerina
// Listener object constructor

> listener http:Listener serviceListener = new(9090);

Reviewer 2: Let's create the listener inside the main as well.

public function main() {
    error? err1 = serviceListener.attach(s, “/basePath”);
    error? err2 = serviceListener.start();
    //...
    error? err3 = serviceListener.gracefulStop();
}

http:Service s = service object {
    resource function get sayHello(http:Caller caller) {}
}
```

### 2.2. Service
Service is a collection of resources functions, which are the network entry points of a ballerina program. 
In addition to that a service can contain public and private functions which can be accessed by calling with `self`.

#### 2.2.1. Service type
```ballerina
public type Service distinct service object {

};
```
Above distinct type is provided by HTTP module and user can include the type as `*http:Service` to refer it.

> The comprehensive typing support is yet to be added to the language

Reviewer 2: ```suggestion
The comprehensive typing support is yet to be added to the language.
```

Reviewer 2: I think better to add a note say until that compiler plugins are used to validate the services.

#### 2.2.2. Service base path

The base path is considered during the request dispatching to discover the service. Identifiers and string literals
are allowed to be stated as base path and it should be started with `/`. The base path is optional and it will be 
default to `/` when not defined. If the base path contains any special charactors, those should be escaped or defined
as string literals

```ballerina

> service http:Service /my/Tes\@tHello/go new http:Listener(9090) {

Reviewer 2: we can remove `http:Service` add something like `hello\-world` for the path.

   resource function get foo() {

   }
}

service http:Service "/Tes@tHello/go" new http:Listener(9090) {
   resource function get foo() {

   }
}
```

A service can be declared in three ways upon the requirement.

#### 2.2.3. Service declaration
The [Service declaration](https://ballerina.io/spec/lang/2021R1/#section_8.3.2) is a syntactic sugar for creating a
service and it is the mostly used approach for creating a service. The declaration gets desugared into creating a 
listener object, creating a service object, attaching the service object to the listener object.

```ballerina
service http:Service /basePath on new http:Listener(9090) {
  resource function get foo() returns string {
      return "hello world";
  }
}
```

#### 2.2.4. Service class declaration

The service value can be instantiated using the service class. This way, user has the completed control of attaching
the service to the listener. The life cycle methods to used to proceed.

```ballerina
service isolated class SClass {
   *http:Service;
   isolated resource function get greeting() returns string {
       return "hello world";
   }
}

listener http:Listener serviceListener = new (9090);

> http:Service httpService = new SClass();

Reviewer 2: We can have the above two inside the main.

public function main() {
   error? err1 = serviceListener.attach(httpService, ["foo", "bar"]);
   error? err2 = serviceListener.'start();
   runtime:registerListener(serviceListener);
}
```

#### 2.2.5. Service constructor expression

```ballerina
listener http:Listener serviceListener = new (9090);

http:Service httpService = @http:ServiceConfig {} service object {

>    resource function get baz() returns string {

Reviewer 2: better to use the same resource in all of them.

and default. If the accessor is unmatched, 405 Method Not Allowed response is returned. When the accessor name is 
stated as default, any HTTP method can be matched to it in the absence of an exact match. Users can define custom 
methods such as copy, move based on their requirement. A resource which can handle any method would look like as 
follows. This is useful when handling unmatched verbs.

```ballerina
resource function 'default NAME_TEMPLATE () {
    
}
```
#### 2.3.2. Resource name
The resource-name represents the path of the resource which is considered during the request dispatching. The name can 
be hierarchical(foo/bar/baz). Each path identifier should be separated by `/` and first path identifier should not 
contain a prefixing `/`. If the paths are unmatched, 404 NOT FOUND response is returned.
```ballerina
resource function post hello() {
    
}
```
Only the identifiers can be used as resource path not string literals. Dot identifier is 
used to denote the `/` only if the path contains a single identifier. 
```ballerina
resource function post .() {
    
}
```
Any special characters can be used in the path by escaping.
```ballerina

> resource function post test\@hello() {

Reviewer 2: `hello\-world` would be better.

outbound response payload. Internally an `http:Response` is created including the given payload value

```ballerina
string|xml|json|byte[]|int|float|decimal|boolean|map<json>|table<map<json>>|(map<json>|table<map<json>>)[]|
mime:Entity[]|stream<byte[], io:Error?>|()
```

Based on the payload types respective header value is added as the `Content-type` of the `http:Response`.

 Type|Content Type
---|---
() | -
string | text/plain
xml | application/xml
byte[], stream<byte[], io:Error?> | application/octet-stream
int, float, decimal, boolean | application/json
map\<json\>, table<map\<json\>>, map\<json\>[], table<map\<json\>>)[] | application/json

The HTTP compiler extension checks the argument of the `respond()` method if the matching payload type is passed as
denoted in the CallerInfo annotation

```ballerina
resource function post foo(@http:CallerInfo {respondType:Person}  http:Caller hc) {
    Person p = {};
    hc->respond(p);
}
```
##### 2.3.4.2. http:Request

The `http:Request` represents the message which came along over the network that modelled as an object along with 

> headers and payload. Listener passes it to the resource function as an argument to be accessed by the user based on 

Reviewer 2: I think the first sentence can be changed to something like the below.

The `http:Request` represents the request which is sent and received over the network which includes headers and the entity body.

            * 8.1.3.3. [Execution order of interceptors](#8133-execution-order-of-interceptors)
        * 8.1.4. [Data binding](#814-data-binding)
        * 8.1.5. [Return to respond](#815-return-to-respond)
     * 8.2. [Error handling](#82-error-handling)
        * 8.2.1. [Trace log](#821-trace-log)
        * 8.2.2. [Access log](#822-access-log)
9. [Security](#9-security)
    * 9.1. [Authentication and Authorization](#91-authentication-and-authorization)
        * 9.1.1. [Declarative Approach](#911-declarative-approach)
            * 9.1.1.1. [Listener - File User Store](#9111-listener---file-user-store)
            * 9.1.1.2. [Listener - LDAP User Store](#9112-listener---ldap-user-store)
            * 9.1.1.3. [Listener - JWT](#9113-listener---jwt)
            * 9.1.1.4. [Listener - OAuth2](#9114-listener---oauth2)
            * 9.1.1.5. [Client - Basic Auth](#9115-client---basic-auth)
            * 9.1.1.6. [Client - Bearer Token Auth](#9116-client---bearer-token-auth)
            * 9.1.1.7. [Client - Self Signed JWT Auth](#9117-client---self-signed-jwt)
            * 9.1.1.8. [Client - Bearer Token OAuth2](#9118-client---bearer-token-oauth2)

>             * 9.1.1.9. [Client - Grant types OAuth2](#9119-client---grant-types-oauth2)

Reviewer 3: ```suggestion
            * 9.1.1.9. [Client - Grant Types OAuth2](#9119-client---grant-types-oauth2)
```

        * 9.1.2 [Imperative Approach](#912-imperative-approach)
            * 9.1.2.1. [Listener - File User Store](#9121-listener---file-user-store)
            * 9.1.2.2. [Listener - LDAP User Store](#9122-listener---ldap-user-store)
            * 9.1.2.3. [Listener - JWT](#9123-listener---jwt)
            * 9.1.2.4. [Listener - OAuth2](#9124-listener---oauth2)
            * 9.1.2.5. [Client - Basic Auth](#9125-client---basic-auth)
            * 9.1.2.6. [Client - Bearer Token Auth](#9126-client---bearer-token-auth)
            * 9.1.2.7. [Client - Self Signed JWT Auth](#9127-self-signed-jwt)

>             * 9.1.2.8. [Client - Bearer token OAuth2](#9128-client---bearer-token-oauth2)
>             * 9.1.2.9. [Client - Grant types OAuth2](#9129-client---grant-types-oauth2)

Reviewer 3: ```suggestion
            * 9.1.2.8. [Client - Bearer Token OAuth2](#9128-client---bearer-token-oauth2)
            * 9.1.2.9. [Client - Grant Types OAuth2](#9129-client---grant-types-oauth2)
```

provide a set of configurations and do not need to be worried more about how authentication and authorization works. 
The user does not have full control over the configuration-driven approach.

The service configurations are used to define the authentication and authorization configurations. Users can 
configure the configurations needed for different authentication schemes and configurations needed for 
authorizations of each authentication scheme. Also, the configurations can be provided at the service level. 
The priority will be given from bottom to top. Then, the auth handler creation and request 
authentication/authorization is handled internally without user intervention. The requests that succeeded both 
authentication and/or authorization phases according to the configurations will be passed to the business logic layer.

##### 9.1.1.1 Listener - File User Store

```ballerina
@http:ServiceConfig {
    auth: [
        {
            fileUserStoreConfig: {},
            scopes: ["admin"]
        }
    ]
}
service /foo on new http:Listener(9090) {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

> ```ballerina

Reviewer 3: ```suggestion

```ballerina
```

                groupNameAttribute: "cn",
                groupNameSearchFilter: "(&(objectClass=groupOfNames)(cn=?))",
                groupNameListFilter: "(objectClass=groupOfNames)",
                membershipAttribute: "member",
                userRolesCacheEnabled: true,
                connectionPoolingEnabled: false,
                connectionTimeout: 5,
                readTimeout: 60
            },
            scopes: ["admin"]
        }
    ]
}
service /foo on new http:Listener(9090) {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

##### 9.1.1.3 Listener - JWT

```ballerina
@http:ServiceConfig {
    auth: [
        {
            jwtValidatorConfig: {
                issuer: "wso2",
                audience: "ballerina",
                signatureConfig: {
                    certFile: "/path/to/public.crt"
                },
                scopeKey: "scp"
            },
            scopes: ["admin"]
        }
    ]
}
service /foo on new http:Listener(9090) {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

##### 9.1.1.4 Listener - OAuth2

> ```ballerina

Reviewer 3: ```suggestion

```ballerina
```

@http:ServiceConfig {
    auth: [
        {
            oauth2IntrospectionConfig: {
                url: "https://localhost:9445/oauth2/introspect",
                tokenTypeHint: "access_token",
                scopeKey: "scp",
                clientConfig: {
                    customHeaders: {"Authorization": "Basic YWRtaW46YWRtaW4="},
                    secureSocket: {
                        cert: "/path/to/public.crt"
                    }
                }
            },
            scopes: ["admin"]
        }
    ]
}
service /foo on new http:Listener(9090) {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

##### 9.1.1.5 Client - Basic Auth

> ```ballerina

Reviewer 3: ```suggestion

```ballerina
```

http:Client c = check new ("https://localhost:9090",
    auth = {
        username: "tom",
        password: "123"
    }
);

public function main() returns error? {
    json response = check c->get("/foo/bar");
    // evaluate response
}
```

##### 9.1.1.6 Client - Bearer Token Auth

> ```ballerina

Reviewer 3: ```suggestion

```ballerina
```

http:Client c = check new ("https://localhost:9090",
    auth = {
        token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiw" +
               "ibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2Q" +
               "T4fwpMeJf36POk6yJV_adQssw5c"
    }
);

public function main() returns error? {
    json response = check c->get("/foo/bar");
    // evaluate response
}
```

##### 9.1.1.7 Client - Self-Signed JWT

> ```ballerina

Reviewer 3: ```suggestion

```ballerina
```

    json response = check c->get("/foo/bar");
    // evaluate response
}
```

##### 9.1.2.8. Client - Bearer Token OAuth2
```ballerina
http:BearerTokenConfig config = {
    token: "JhbGciOiJIIiwiaWF0IjUzI1NiIsInR5cCI6IkpXVCJ9WIiOiIxMjM0NTY3ODkwI"
};
http:ClientBearerTokenAuthHandler handler = new (config);

http:Client c = check new ("https://localhost:9090");

public function main() returns error? {
    http:Request req = new;
    req = check handler.enrich(req);
    json response = check c->get("/foo/bar");
    // evaluate response
}
```

##### 9.1.2.9. Client - Grant Types OAuth2
```ballerina
http:OAuth2ClientCredentialsGrantConfig config = {
    tokenUrl: "https://localhost:8080/oauth2/token/authorize",
    clientId: "3MVG9YDQS5WtC11paU2WcQjBB3L5w4gz52uriT8ksZ3nUVjKvrfQMrU4uvZohTftxS",
    clientSecret: "9205371918321623741"
};
http:ClientOAuth2Handler handler = new (config);

http:Client c = check new ("https://localhost:9090");

public function main() returns error? {
    http:Request req = new;
    req = check handler->enrich(req);
    json response = check c->get("/foo/bar");
    // evaluate response
}
```

### 9.2 SSL/TLS and Mutual SSL

> The HTTPS listener could connect to or interact with an HTTPS client. The http:ListenerSecureSocket configuration

Reviewer 3: ```suggestion
The HTTPS listener could connect to or interact with an HTTPS client. The `http:ListenerSecureSocket` configuration
```

of the listener exposes the HTTPS connection related configs.

#### 9.2.1 Listener - SSL/TLS
```ballerina
listener http:Listener securedEP = new(9090,
    secureSocket = {
        key: {

>             certFile: "../resource/path/to/public.crt",
>             keyFile: "../resource/path/to/private.key"

Reviewer 3: ```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

        }
    }
);

service /foo on securedEP {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

#### 9.2.2 Listener - Mutual SSL

The mutual SSL support which is a certificate-based authentication process in which two parties 
(the client and server) authenticate each other by verifying the digital certificates. It ensures that both 
parties are assured of each other’s identity.

```ballerina
listener http:Listener securedEP = new(9090,
    secureSocket = {
        key: {

>             certFile: "../resource/path/to/public.crt",
>             keyFile: "../resource/path/to/private.key"

Reviewer 3: ```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

        },

        mutualSsl: {
            verifyClient: http:REQUIRE,

>             cert: "../resource/path/to/public.crt"

Reviewer 3: ```suggestion
            cert: "/path/to/public.crt"
```

        },

        protocol: {
            name: http:TLS,
            versions: ["TLSv1.2", "TLSv1.1"]
        },

        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]

    }
);
service /foo on securedEP {
    resource function get bar() returns string {
        return "Hello, World!";
    }
}
```

#### 9.2.3 Client - SSL/TLS

```ballerina
http:Client securedEP = check new("https://localhost:9090",
    secureSocket = {

>         cert: "../resource/path/to/public.crt"

Reviewer 3: ```suggestion
        cert: "/path/to/public.crt"
```

    }
);

public function main() returns error? {
    string response = check securedEP->get("/foo/bar");
    io:println(response);
}
```

#### 9.2.4 Client - Mutual SSL

```ballerina
http:Client securedEP = check new("https://localhost:9090",
    secureSocket = {
        key: {

>             certFile: "../resource/path/to/public.crt",
>             keyFile: "../resource/path/to/private.key"

Reviewer 3: ```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

        },

>         cert: "../resource/path/to/public.crt",

Reviewer 3: ```suggestion
        cert: "/path/to/public.crt",
```

service hello\-world new http:Listener(9090) {
   resource function get foo() {

   }
}

service http:Service "hello-world" new http:Listener(9090) {
   resource function get foo() {

   }
}
```

A service can be declared in three ways upon the requirement.

#### 2.2.3. Service declaration
The [Service declaration](https://ballerina.io/spec/lang/2021R1/#section_8.3.2) is a syntactic sugar for creating a
service and it is the mostly used approach for creating a service. The declaration gets desugared into creating a 
listener object, creating a service object, attaching the service object to the listener object.

```ballerina
service http:Service /foo/bar on new http:Listener(9090) {
  resource function get greeting() returns string {
      return "hello world";
  }
}
```

#### 2.2.4. Service class declaration

The service value can be instantiated using the service class. This way, user has the completed control of attaching
the service to the listener. The life cycle methods to used to proceed.

```ballerina
service isolated class SClass {
   *http:Service;
   resource function get greeting() returns string {
       return "hello world";
   }
}

public function main() {

>    listener http:Listener serviceListener = new (9090);

Reviewer 2: Can we use `listener` keyword inside main ?

with any annotation or additional detail. This parameter is not compulsory and not ordered. The type of query param 
are as follows

```ballerina
type BasicType boolean|int|float|decimal|string|map<json>;
public type QueryParamType ()|BasicType|BasicType[];
```

The same query param can have multiple values. In the presence of multiple such values,  If the user has specified 
the param as an array type, then all values will return. If not the first param values will be returned. As per the 
following resource function, the request may contain at least two query params with the key of bar and id.
Eg : “/hello?bar=hi&id=56”

```ballerina
resource function get hello(string bar, int id) { 
    
}
```

If the query parameter is not defined in the function signature, then the query param binding does not happen. If a 
query param of the request URL has no corresponding parameter in the resource function, then that param is ignored. 
If the parameter is defined in the function, but there is no such query param in the URL, that request will lead 
to a 400 BAD REQUEST error response unless the type is nilable (string?)

> The query param consists of query name and values. Sometimes user may sent query without value(`foo:`). In such

Reviewer 2: ```suggestion
The query param consists of query name and values. Sometimes user may send query without value(`foo:`). In such
```

           "responses": {
              "200": {
                 "description": "Ok",
                 "content": {
                    "text/plain": {
                       "schema": {
                          "type": "string"
                       }
                    }
                 }
              }
           }
        }
     }
  },
  "components": {}
}
```

Output of OPTIONS call to service base path

```ballerina
curl -v localhost:9090/hello -X OPTIONS 
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9090 (#0)
> OPTIONS /hello HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 204 No Content
< allow: GET, OPTIONS
< link: </hello/openapi-doc-dygixywsw>;rel="service-desc"
< server: ballerina/2.0.0-beta.2.1
< date: Thu, 19 Aug 2021 13:47:29 +0530
< 
* Connection #0 to host localhost left intact
* Closing connection 0
```

### 2.4. Client
A client allows the program to send network messages to a remote process according to the HTTP protocol. The fixed 
remote functions of the client object correspond to distinct network operations defined by the HTTP protocol.

> The client init function requires port and configuration to initialize the client. 

Reviewer 2: ```suggestion
The client init function requires a valid URL and optional configuration to initialize the client. 
```

```ballerina
http:Client clientEP = check new ("http://localhost:9090", { httpVersion: "2.0" });
```

#### 2.4.1 Client types
The client configuration can be used to enhance the client behaviour.

```ballerina
public type ClientConfiguration record {|
    string httpVersion = HTTP_1_1;
    ClientHttp1Settings http1Settings = {};
    ClientHttp2Settings http2Settings = {};
    decimal timeout = 60;
    string forwarded = "disable";
    FollowRedirects? followRedirects = ();
    PoolConfiguration? poolConfig = ();
    CacheConfig cache = {};
    Compression compression = COMPRESSION_AUTO;
    ClientAuthConfig? auth = ();
    CircuitBreakerConfig? circuitBreaker = ();
    RetryConfig? retryConfig = ();
    CookieConfig? cookieConfig = ();
    ResponseLimitConfigs responseLimits = {};
    ClientSecureSocket? secureSocket = ();
|};
```

Based on the config, the client object will be accompanied by following client behaviours. Following clients cannot be
instantiated calling `new`, instead user have to enable the config in the `ClientConfiguration`.

> ##### 2.4.1.1 Secure client 

Reviewer 2: I wonder if the topics should be as follows,

2.4.1.1 Security
2.4.1.2 Caching
2.4.1.3 Redirect

Provides secure HTTP remote functions for interacting with HTTP endpoints. This will make use of the authentication
schemes configured in the HTTP client endpoint to secure the HTTP requests.
```ballerina
http:Client clientEP = check new("https://localhost:9090",
    auth = {
        username: username,
        password: password
    },
    secureSocket = {
        cert: {
            path: TRUSTSTORE_PATH,
            password: "ballerina"
        }
    }
);
```

##### 2.4.1.2 Caching client

> An HTTP caching client implementation which wraps with an HTTP caching layer once `cache` config is enabled.

Reviewer 2: We don't need to talk about implementation details in the spec.

HTTP request method. Also `execute` remote function is useful to send request with custom HTTP verbs such as `move`, 
`copy`, ..etc.

```ballerina
# Invokes an HTTP call with the specified HTTP verb.
remote isolated function execute(string httpVerb,  string path, RequestMessage message, 
        map<string|string[]>? headers = (), string? mediaType = (), TargetType targetType = <>)
        returns targetType|ClientError;

# The forward() function can be used to invoke an HTTP call with inbound request's HTTP verb
remote isolated function forward(string path, Request request, TargetType targetType = <>)
        returns  targetType|ClientError;
```

###### 2.4.2.4 HTTP2 additional methods
Following are the HTTP2 client related additional remote functions to deal with promises and responses.

```ballerina

# Submits an HTTP request to a service with the specified HTTP verb.
# The submit() function does not give out a http:Response as the result.
# Rather it returns an http:HttpFuture which can be used to do further interactions with the endpoint.
remote isolated function submit(string httpVerb, string path, RequestMessage message)
    returns HttpFuture|ClientError;

> # This just pass the request to actual network call.

Reviewer 2: Instead of `This just pass the request to actual network call.` maybe we can say `Passes the request to actual network call.`

remote isolated function  getResponse(HttpFuture httpFuture) returns Response|ClientError;

# This just pass the request to actual network call.

> remote isolated function  hasPromise(HttpFuture httpFuture) returns boolean;

Reviewer 2: Seems like there are additional spaces.

}
```

### 4.3. Payload annotation
The payload annotation has two usages. It is used to decorate the resource function payload parameter and to decorate 
the resource return type. 

```ballerina
public type Payload record {|
    string|string[] mediaType?;
|}
```

#### 4.3.1. Payload binding parameter

The request payload binding is supported in resource functions where users can access it through a resource function 
parameter. The @http:Payload annotation is specially introduced to distinguish the request payload with other 
resource function parameters. The annotation can be used to specify values such as mediaType...etc. Users can 
define the potential request payload content type as the mediaType to perform some pre-validations as same as 
Consumes resource config field.

```ballerina
resource function post hello(@http:Payload {mediaType:["application/json", "application/ld+json"]} json payload)  {
    
}
```

During the runtime, the request content-type header is matched against the mediaType field value to validate. If the 
validation fails, the listener returns an error response with the status code of 415 Unsupported Media Type. 
Otherwise the dispatching moves forward. 

> #### 4.3.2. anydata return value info

Reviewer 2: ```suggestion
#### 4.3.2. Anydata return value info
```

        cert: "/path/to/public.crt",
        protocol: {
            name: http:TLS
        },
        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]

    }
);
public function main() returns error? {
    string response = check securedEP->get("/foo/bar");
    io:println(response);
}
```

## 10. Protocol upgrade
### 10.1. HTTP/2
The version 2 of HTTP protocol is supported in both Listener and Client space which could be configured through the 
respective configuration.

```ballerina
// Listener declaration
listener http:Listener http2ServiceEP = new (7090, config = {httpVersion: "2.0"});

// Client declaration
http:Client clientEP = check new ("http://localhost:7090", {httpVersion: "2.0"});
```

There are few API level additions when it comes to the HTTP/2 design such as Push promise and promise response.
#### 10.1.1. Push Promise and Promise Response
Push Promise and Promise response are the only application level new semantics which are introduced by HTTP2.

Other protocol changes such as streams, messages, frames, request prioritization, flow control, header compression, 
etc. are all lower level changes that can be handled by the HTTP listener seamlessly from the user.

> ### 10.2. Websocket

Reviewer 2: I think this is not needed. As decoupled HTTP from Websocket

resource function.

```ballerina
service class RequestInterceptor {
   *http:RequestInterceptor;
 
   resource function 'default [string… path](http:FilterContext ctx, http:Caller caller,
                       http:Request req) returns error? {
       request.setHeader("X-requestHeader", "RequestInterceptor");
       ctx.next();
   }
}
```

Since interceptors work with network activities, it must be either a remote or resource function. In this case resource 
functions are used for RequestInterceptor as it gives more flexibility. With resource functions interceptors can be engaged 
based on HTTP method and path.

For instance consider a scenario where there are two resources; one on path foo whereas the other on path bar. If the 
user writes a interceptor as follows, it would only get hit when the request is directed to foo resource.

```ballerina
service class RequestInterceptor {
   *http:RequestInterceptor;
 
   resource function 'default foo(http:FilterContext ctx, http:Caller caller,
                       http:Request req) returns error? {
       request.setHeader("X-requestHeader", "RequestInterceptor");
       ctx.next();
   }
}
```

> ##### 8.1.1.1 FilterContext  

Reviewer 4: Don't we use the `RequestContext` term rather than `FilterContext`?

Following is the rough definition of the interceptor context.
```ballerina
public isolated class FilterContext {
    private final map<value:Cloneable|isolated object {}> attributes = {};

    public isolated function add(string 'key, value:Cloneable|isolated object {} value) {
        if value is value:Cloneable {
            lock {
                self.attributes['key] = value.clone();
            }
        }
        else {
            lock {
                self.attributes['key] = value;
            }   
        }
    }

    public isolated function get(string 'key) returns value:Cloneable|isolated object {} {
        lock {
            return self.attributes.get('key);
        }
    }

    public isolated function remove(string 'key) {
        lock {
            value:Cloneable|isolated object {} _ = self.attributes.remove('key);
        }
    }

>     public isolated function next() returns error? = external;

Reviewer 4: `FilterContext`->`next()` should return `NextService|error?` right?

}
```

##### 8.1.1.2 next() Method  
However, there is an addition when it comes to FilterContext. A new method namely, next() is introduced to control 
the execution flow. Users must invoke next() method in order to trigger the next interceptor in the pipeline. 
Previously, this was controlled by returning a boolean value which is quite cryptic and confusing.

Moreover, invoking next() method results in immediately executing the next interceptor in the pipeline. Which means 
whatever below the next() method will only get executed after the completion of execution of the next Interceptor.

If the users forget to call next() and don't respond either, the request will be left hanging.

##### 8.1.1.3 Returning error?

> Resource functions can return only error?. In the case of an error, interceptor chain execution jumps to the last 

Reviewer 4: Resource functions can return `NextService|error?`. They can return only `error?` **when using the `Caller`**. 
Also, in case of an error, interceptor pipeline execution jumps to the **nearest `RequestErrorInterceptor`** in the pipeline. This `RequestErrorInterceptor` is not necessarily the last interceptor in the pipeline, it can be anywhere in the chain.

of request or response interceptor chain. Meaning these interceptors are only executed as the last interceptor in the chain. The 
framework automatically adds default RequestErrorInterceptor and ResponseErrorInterceptor which basically prints the error 
message to the console.

Users can override these interceptors by defining their own ones as follows. Users don’t have to specifically engage 
these interceptors as they only have fixed positions and they are always executed. The only additional argument in this 
case is error err.

```ballerina
service class RequestErrorInterceptor {
   *http:RequestErrorInterceptor;
 
   remote function 'default [string… path](http:FilterContext ctx, http:Caller caller,
                       http:Request req, error err) returns error? {
       // deal with the error
   }
}
```

#### 8.1.3 Engaging Interceptors
##### 8.1.3.1 Service Level
Interceptors could get engaged at service level. One reason for this is that users may 
want to engage two different interceptor chains for each service even though it is attached to the same Listener. At the 

> listener level resource function paths are relative to the service base path.

Reviewer 4: At the **service** level

```ballerina
@http:ServiceConfig{
   interceptors: [requestInterceptor, responseInterceptor]
}
```

##### 8.1.3.2 Listener Level
Interceptors could get engaged at Listener level as well. At the listener level resource function paths are 
relative to the /.
```ballerina
listener http:Listener echoListener = new http:Listener(9090, config = {interceptors: [requestInterceptor, responseInterceptor]});
```

##### 8.1.3.3 Execution Order of Interceptors

> These interceptors are always there in the pipeline and executed as the last.

Reviewer 4: This line is not related to the topic

![img.png](img.png)
In the above example blue dashed box represents the RequestErrorInterceptor and blue boxes simply represent the 
RequestInterceptors whereas green dashed box represents the ResponseErrorInterceptor(TBA) and green boxes simply represent the 
ResponseInterceptors. The new execution orders is as follows,
```
RequestInterceptor/RequestErrorInterceptor: 1, 2, 4, 5
ResponseInterceptor/ResponseErrorInterceptor: 5, 3, 0
```

> 

Reviewer 4: Do we need to say about the execution order when interceptors are engaged at both service and listener levels?

`ListenerLevel : RequestInterceptors` -> `ServiceLevel : RequestInterceptors` -> `TargetService` -> `ServiceLevel : ResponseInterceptors` -> `ListenerLevel : ResponseInterceptors`
